File: java_solutions/20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
// semicolons : 1
class Solution {
    public boolean isValid(String s) {
        return Optional.of(s.chars().mapToObj(c -> (char)c).reduce(new Object[]{new Stack<Character>(), true}, (state, c) -> Optional.of((boolean)state[1]).filter(v -> v).map(v -> Optional.ofNullable(Map.of('(', ')', '[', ']', '{', '}').get(c)).map(close -> new Object[]{Stream.of((Stack<Character>)state[0]).peek(st -> st.push(close)).findFirst().get(), true}).orElseGet(() -> Optional.of((Stack<Character>)state[0]).filter(st -> !st.isEmpty() && st.pop() == c).map(st -> state).orElse(new Object[]{state[0], false}))).orElse(state), (a, b) -> a)).filter(state -> Optional.of((boolean)state[1]).filter(v -> v).filter(v -> ((Stack<Character>)state[0]).isEmpty()).isPresent()).isPresent();
    }
}
// @lc code=end

File: kotlin_solutions/20.valid-parentheses.kt
/*
 * @lc app=leetcode id=20 lang=kotlin
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution { fun isValid(s: String): Boolean = s.fold(mutableListOf<Char>()) { stack, c -> when (c) { '(', '{', '[' -> stack.apply { add(c) } else -> stack.takeIf { it.isNotEmpty() && mapOf(')' to '(', '}' to '{', ']' to '[')[c] == it.last() } ?.apply { removeAt(size - 1) } ?: return false } }.isEmpty() }
// @lc code=end

